### ISSUE TITLE
URL specific operations

### ISSUE THREAD

**Issue by Virviil** [ericmj/xhttp#42]
Elixir Core seems not to have library to work with URLs, including last specifications with IDNA, punycodes and all related features.

[Hackney](https://hex.pm/packages/hackney) utilizes [IDNA](https://hex.pm/packages/idna) package for this purposes.

Is there a plan to implement something similar in Elixir, may be to use in XHTTP, or even include it into Elixir Core?

For example, Node.JS has [URL](https://nodejs.org/api/url.html) API inside it's STL.

If it will be valuable, I can try to implement URL related parts of [WHATWG specs](https://url.spec.whatwg.org/) as a part of this project, or as a separate project, if it's better.

**Comment by whatyouhide** [comment]
Forgive my ignorance but I am not an expert in this area. What would be missing from the Elixir URI module in the stdlib that we might want to add here?

**Comment by Virviil** [comment]
Well, I'm not sure if it's needed in Elixir Core, but for HTTP Client next functionality might be useful:

1. URL normalization:
    ```
    # all examples are working in modern browsers
    "https://////www.google.com" -> "https:/www.google.com"
    "https://你好你好" -> "https://xn--6qqa088eba/"
    "https://０Ｘｃ０．０２５０．０１" -> "https://192.168.0.1/"
    ```

2. URL validation:

    ```
    https://www .google.com -> false
    ```

3. Relative URLs:

    ```
    "../foo?a=bar#bazz" with base "https://www.google.com/maps" -> "https://www.google.com/foo?a=bar#bazz"
    ```

**Comment by Virviil** [comment]
@ericmj This sounds really good. We will pend with investigations until something will become more clear. Hope this thread will help with coordinations upon given topic in the future.

**Comment by chulkilee** [comment]
I think *this* HTTP library should focus on HTTP handling, rather than having all convenient features for HTTP clients.

**For "url helper methods", such as validation, auto-fix:**

* "modern" browser behavior is not necessarily "standard"
* If bad value is given, then HTTP library should fail, instead of trying to auto-fix it

Those "convenience" features should be built on higher level, not this HTTP library level. For example, I'm using [tesla](https://hex.pm/packages/tesla) and I can do all with Tesla middleware (e.g. see [Tesla.Middleware.BaseUrl](https://hexdocs.pm/tesla/Tesla.Middleware.BaseUrl.html)) - so that underlying HTTP library (e.g. hackney, httpc..) doesn't need to handle this.

**For IDNA:**

We may end up with common interface for IDNA (like we do for Json with jason and poison) so that we can evolve/release mint without being coupled with specific idna library.

**Comment by whatyouhide** [comment]
I agree with @chulkilee. I think we can close this issue and open one that will remind us to look into a IDNA interface. Wdyt @ericmj?

**Comment by ericmj** [comment]
I agree. Please open a separate issue. I think having a separate library that handles IDNA would be ideal, we would only need some documentation on how to use the the library with Mint.
